python/pyspark/ml/torch/deepspeed/deepspeed_distributer.py

```diff
@@ -0,0 +1,167 @@
+        local_mode: bool = True,
+        use_gpu: bool = True,
+        deepspeed_config=None,
+    ):
+        super().__init__(num_processes, local_mode, use_gpu)
+        self.deepspeed_config = deepspeed_config
+        self.temp_deepspeed_fname = None
+        self.input_params = self._create_input_params()
+        self.worker_hosts = self._setup_hostfile_info()
+        self.setup_env()
+
+    def _get_gpus_on_node(self, executor_ip: str):
+        # TODO: ask Ricky, Lu, or Maddie if this is the best way to get the GPU information of a particular worker node
+        command = f"ssh {executor_ip} nvidia-smi -L | grep GPU | wc -l"  # pyspark doesn't support this out of the box for some reason, so sadge
+        proc_res = subprocess.run(command, capture_output=True, text=True, shell=True)
+        if proc_res.returncode:
+            raise RuntimeError(
+                f"something went wrong when running the command {command}. Is nvidia-smi installed?"
+            )
+        num_gpus_on_worker = proc_res.stdout
+        return int(num_gpus_on_worker)
+
+    def _assign_procs_to_worker(self, gpu_node_map: Dict[str, int]) -> Dict[str, int]:
+        procs_left = self.num_processes
+        workers_left_to_serve = len(gpu_node_map)
+        average_procs_per_node = procs_left // workers_left_to_serve
+        gpu_mapped_to_node = {}
+        # sorting allows us to just do a single pass, as filling the smallest capacity nodes first will allow for a single pass
+        sorted_buckets = sorted(gpu_node_map.items(), key=lambda x: x[1])
+        for worker, capacity in sorted_buckets:
+            average_procs_per_node = procs_left // workers_left_to_serve
+            gpu_mapped_to_node[worker] = min(average_procs_per_node, capacity)
+            procs_left -= gpu_mapped_to_node[worker]
+            workers_left_to_serve -= 1
+        if procs_left != 0:
+            self.logger.warning(
+                msg=f"There are not enough GPUS to fully assign processes to nodes; there are {procs_left} processes left over"
+            )
+        return gpu_mapped_to_node
+
+    def _setup_hostfile_info(self):
+        worker_hosts = [
+            executor.host()
+            for executor in self.spark.sparkContext._jsc.sc().statusTracker().getExecutorInfos()
+        ]  # we should check if this returns the driver or not
+        worker_count = len(worker_hosts)  # find out if this number includes the driver or not
+        rdd = spark.sparkContext.parallelize(range(worker_count), numSlices=worker_count)
+
+        # what do I do if the use_gpu flag is false?
+        slots_on_workers = {}
+        if self.use_gpu:
+            for worker_host in worker_hosts:
+                slots_on_workers[worker_host] = self._get_gpus_on_node(worker_host)
+        else:
+            raise RuntimeError("Deepspeed doesn't work with non-GPU clusters at this time")
+
+        assigned_slots = self._assign_procs_to_worker(slots_on_workers)
```

Comment:
It doesn't seem intuitive for users to pass in the total `num_processes` and let `DeepspeedDistributor` assign them to the GPUs of each node. Can it just be `f"{worker_host} slots={self._get_gpus_on_node(worker_host)}"`?
https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node